proxy/src/telemetry/metrics/process.rs

```diff
@@ -0,0 +1,153 @@
+use std::fmt;
+use super::{Counter, Gauge, Metric};
+
+pub use self::imp::Sensor;
+
+#[derive(Copy, Clone)]
+pub struct ProcessMetrics {
+    cpu_seconds_total: Counter,
+    open_fds: Gauge,
+    max_fds: Option<Gauge>,
+    virtual_memory_bytes: Gauge,
+    resident_memory_bytes: Gauge,
+}
+
+impl ProcessMetrics {
+    metrics! {
+        process_cpu_seconds_total: Counter {
+            "Total user and system CPU time spent in seconds."
+        },
+        process_open_fds: Gauge { "Number of open file descriptors." },
+        process_max_fds: Gauge { "Maximum number of open file descriptors." },
+        process_virtual_memory_bytes: Gauge {
+            "Virtual memory size in bytes."
+        },
+        process_resident_memory_bytes: Gauge {
+            "Resident memory size in bytes."
+        }
+    }
+}
+
+impl fmt::Display for ProcessMetrics {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Self::process_cpu_seconds_total.fmt_help(f)?;
+        Self::process_cpu_seconds_total.fmt_metric(
+            f,
+            self.cpu_seconds_total
+        )?;
+
+        Self::process_open_fds.fmt_help(f)?;
+        Self::process_open_fds.fmt_metric(f, self.open_fds)?;
+
+        if let Some(ref max_fds) = self.max_fds {
+            Self::process_max_fds.fmt_help(f)?;
+            Self::process_max_fds.fmt_metric(f, *max_fds)?;
+        }
+
+        Self::process_virtual_memory_bytes.fmt_help(f)?;
+        Self::process_virtual_memory_bytes.fmt_metric(
+            f,
+            self.virtual_memory_bytes
+        )?;
+
+        Self::process_resident_memory_bytes.fmt_help(f)?;
+        Self::process_resident_memory_bytes.fmt_metric(
+            f,
+            self.resident_memory_bytes
+        )
+    }
+}
+
+#[cfg(target_os = "linux")]
+mod imp {
+    use super::*;
+    use super::super::{Counter, Gauge};
+
+    use std::{io, fs};
+
+    use procinfo::pid;
+    use libc::{self, pid_t};
+
+    #[derive(Debug)]
+    pub struct Sensor {
+        page_size: usize,
+    }
+
+    impl Sensor {
+        pub fn new() -> io::Result<Sensor> {
+            let page_size = match unsafe { libc::sysconf(libc::_SC_PAGESIZE) } {
+                e if e < 0 => {
+                    let error = io::Error::last_os_error();
+                    error!("error getting page size: {:?}", error);
+                    return Err(error);
+                },
+                page_size => page_size as usize,
+            };
+            Ok(Sensor {
+                page_size,
+            })
+        }
+
+        pub fn metrics(&self) -> io::Result<ProcessMetrics> {
+            let stat = pid::stat_self()?;
```

Comment:
If I understand correctly:
- this has to interact with blocking system calls
- and this is run each time the metrics endpoint is accessed.

If this is the case, we may need to take some care here. My first inclination is that this seems like a good candidate for the new tokio `blocking` api (cc @carllerche ), though I can see how that might make this API more difficult, potentaily.

I don't think this is a hard blocker, but I'm curious if you have any thoughts about how to manage the case where this call is either very slow or hangs indefinitely.